tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/TestRunner.cs

```diff
@@ -60,6 +58,172 @@ public void RunTestDefinition(JsonDrivenTestCase testCase)
+                                        unsupportedCompressors.Add(bsonCompressorType);
+                                    }
+                                }
+                                unsupportedCompressors.ForEach(c => expectedBsonArray.Remove(c));
+                                expectedBsonArray.Should().Be(bsonArray);
+                            }
+                            break;
+                        case "connecttimeoutms":
+                            AssertTimeStamp(connectionString.ConnectTimeout, expectedOption.Value);
+                            break;
+                        case "heartbeatfrequencyms":
+                            AssertTimeStamp(connectionString.HeartbeatInterval, expectedOption.Value);
+                            break;
+                        case "journal":
+                            AssertBoolean(connectionString.Journal, expectedOption.Value);
+                            break;
+                        case "localthresholdms":
+                            AssertTimeStamp(connectionString.LocalThreshold, expectedOption.Value);
+                            break;
+                        case "maxidletimems":
+                            AssertTimeStamp(connectionString.MaxIdleTime, expectedOption.Value);
+                            break;
+                        case "maxstalenessseconds":
+                            AssertTimeStamp(connectionString.MaxStaleness, expectedOption.Value, false);
+                            break;
+                        case "readconcernlevel":
+                            AssertEnum(connectionString.ReadConcernLevel, expectedOption.Value);
+                            break;
+                        case "readpreference":
+                            AssertEnum(connectionString.ReadPreference, expectedOption.Value);
+                            break;
+                        case "readpreferencetags":
+                            {
+                                var bsonArray = new BsonArray();
+                                var tags = connectionString.ReadPreferenceTags;
+                                foreach (var tag in tags)
+                                {
+                                    if (!tag.IsEmpty)
+                                    {
+                                        var bsonDocument = new BsonDocument();
+                                        foreach (var innerTag in tag.Tags)
+                                        {
+                                            bsonDocument.Add(innerTag.Name, innerTag.Value);
+                                        }
+                                        bsonArray.Add(new BsonDocument(bsonDocument));
+                                    }
+                                }
+                                expectedOption.Value.Should().Be(bsonArray);
+                            }
+                            break;
```

Comment:
Refactor this case to:

    case "readpreferencetags":
        var readPreferenceTags = ConvertReadPreferenceTagsToBsonArray(connectionString.ReadPreferenceTags);
        readPreferenceTags.Should().Be(expectedOption.Value.AsBsonArray);
        break;

with helper method:

    private BsonArray ConvertReadPreferenceTagsToBsonArray(IReadOnlyList<TagSet> tagSets)
    {
        var tagSetsArray = new BsonArray();
        foreach (var tagSet in tagSets)
        {
            if (!tagSet.IsEmpty)
            {
                var tagSetDocument = new BsonDocument();
                foreach (var tag in tagSet.Tags)
                {
                    tagSetDocument.Add(tag.Name, tag.Value);
                }
                tagSetsArray.Add(tagSetDocument);
            }
        }
        return tagSetsArray;
    }